### ISSUE TITLE
Modify 'import css' order more than 2 times, it will not work

### ISSUE THREAD

**Issue by shield2018** [webpack-contrib/mini-css-extract-plugin#406]
* Operating System:mac os 10.14.4 
* Node Version:8.9.0
* NPM Version:6.8.0
* webpack Version:4.16.5
* mini-css-extract-plugin Version:0.7.0

### Expected Behavior
when I use webpack-dev-server. I modify css order. The out css should be modified at the same time. 
For example:
```
import 'a.less';
import 'b.less';
```
out:
```
.a{
}
.b{
}
```
modify:
```
import 'b.less';
import 'a.less';
```
out:
```
.b{
}
.a{
}
```

### Actual Behavior
Only the first modification works. Don't work from the second time.
For example:
```
import 'a.less';
import 'b.less';
```
out:
```
.a{
}
.b{
}
```

modify:
```
import 'b.less';
import 'a.less';
```
out:
```
.b{
}
.a{
}
```
modify:
```
import 'a.less';
import 'b.less';
```
out:
```
.b{
}
.a{
}
```
not change

### How Do We Reproduce?
Modify import more than 2 times

**Comment by shield2018** [comment]
this is because CssModule does not have 'libIdent' method.

**Comment by evilebottnawi** [comment]
Please create minimum reproducible test repo, thanks

**Comment by evilebottnawi** [comment]
Looks like problem in `less-loader`

**Comment by shield2018** [comment]
I reanalyzed the reason and found in `this.sortItemsWithModuleIds();`. the modules in chunks was sorted by `sortModuleById`. it will Returns 0 when the number is compared to a string and Different order of characters and Numbers can lead to different results. 
Although `import` changed the order ,but after 'sortModuleById' it same as last time. `renderContentAsset` will not run because the cache is turned on.

**Comment by shield2018** [comment]
example
```
var item1 = [0,'./src/pages/index/index.js',3,'./src/pages/index/index.less',1,2,'./c','./Users','./Users',"./Users", "./Users", "./Users","./Users", 4,'./node_modules'];
// var item2 = ['./src/pages/index/index.js',6,7,'./src/pages/index/index.less',1,'./c',2,'./Users',3,'./Users',4,"./Users", "./Users", "./Users","./Users", 8,'./node_modules'];
var item2 = [0,'./src/pages/index/index.js',1,'./src/pages/index/index.less',2,3,'./c','./Users','./Users',"./Users", "./Users", "./Users","./Users", 4,'./node_modules'];

item1.sort((a,b)=>{
    if (a < b) return -1;
     if (b < a) return 1;
     return 0;
})
item2.sort((a,b)=>{
    if (a < b) return -1;
     if (b < a) return 1;
     return 0;
})
console.log(item1);
console.log(item2);
```
item1 is the same as item2.

**Comment by shield2018** [comment]
node 12.3.1 result is right

**Comment by evilebottnawi** [comment]
Please create **minimum reproducible test repo**

**Comment by shield2018** [comment]
repo: https://github.com/shield2018/mini-css-extract-plugin-406-demo

**Comment by evilebottnawi** [comment]
Thanks looks like bug

**Comment by sokra** [comment]
I think this was already fixed by https://github.com/webpack/webpack/pull/8243

Could you try upgrading to the latest webpack version?

**Comment by sokra** [comment]
No doesn't seem to be fixed

**Comment by sokra** [comment]
Ok I found the issue.

While the correct order is determined when emitting the asset, calculating the content hash is not done in with the correct order:

https://github.com/webpack-contrib/mini-css-extract-plugin/blob/392c4ae68c9384230d6652e7fee277a702deacd7/src/index.js#L269-L273

I was able to fix the problem after changing this to:

``` js
        const modules = Array.from(chunk.modulesIterable).filter(module => module.type === MODULE_TYPE);
        const orderedModules = this.determineModulesInOrder(chunk, modules);

        for (const m of orderedModules) {
          if (m.type === MODULE_TYPE) {
            m.updateHash(hash);
          }
        }
```

---

I also extracted the following code snippet into a new function `determineModulesInOrder(chunk, modules, compilation, requestShortener)`:

https://github.com/webpack-contrib/mini-css-extract-plugin/blob/392c4ae68c9384230d6652e7fee277a702deacd7/src/index.js#L454-L555

To avoid emitting warnings in the hashing part I wrapped

https://github.com/webpack-contrib/mini-css-extract-plugin/blob/392c4ae68c9384230d6652e7fee277a702deacd7/src/index.js#L534-L543

in `if(compilation) {}`

---

There are a bunch of other problems:

`this.id = ''`

This doesn't work as well as expected for multiple reasons:

The `NamedModulesPlugin` kind of ignores this while trying to eliminate duplicate ids.

`id`s are cleared when recompiling, which resets this on incremental compilation.

In Chunks Modules are sorted by id, so this kind of affects the order.

As hack we could use this instead:

``` js
  get id() {
    return '';
  }

  set id(value) {}
```

webpack 5 changes a lot of this stuff. I. e. there is no `id` property on Modules anymore and ID assigning is more reliable when modules already have ids.

---

Send a PR

**Comment by shield2018** [comment]
https://github.com/webpack-contrib/mini-css-extract-plugin/blob/392c4ae68c9384230d6652e7fee277a702deacd7/src/index.js#L269-L273

changing this to:
```javascript
const modules = Array.from(chunk.modulesIterable).filter(module => module.type === MODULE_TYPE);
modules.sort((a,b)=>{
  return a.index-b.index;
})
for (const m of modules) {
  if (m.type === MODULE_TYPE) {
    m.updateHash(hash);
  }
}
```

**Comment by alexander-akait** [comment]
Fixed for webpack v5 and v1.3.8
